proxy/proxy.go

```diff
@@ -693,6 +744,10 @@ func (p *Proxy) errorResponse(ctx *context, err error) {
 		ctx.responseWriter.Header().Set("X-Circuit-Open", "true")
 	case err == errRouteLookupFailed:
 		code = p.defaultHTTPStatus
+	case perr.err == errRatelimitError:
```

Comment:
the additionalHeader field on the proxyError type could be used in a more generic way, not only for the rate limit errors, even if currently we only use for that rate limiter response. How about copying all the headers from the additionalHeader with the copyHeader function to the response whenever the error is of type proxyError? (A small bonus refactoring could be to do the same to the circuit breaker error, by moving the above "X-Circuit-Open" header to the additionHeader field where the errCircuitBreakerOpen var is defined.)

Reply:
I like the idea.
I think, I also have to check for "ok" if I access perr.err, which could be a nil ptr dereference, if not ok.